veps/sig-compute/10-dra-devices/vep.md

```diff
@@ -0,0 +1,781 @@
+
+### Unsupported Usecases
+
+1. Devices where the DRA driver does not set the attributes needed to configure libvirt dom XML for the devices will not
+   be supported.
+    1. In the future a standardization could be envisioned for a PCI device where the attributes are set
+       automatically through the DRA framework. When this is achieved, any DRA device should be available in KubeVirt VMs
+2. NetworkDevices or Storage Devices which need a lifecycle of its own will not be supported through HostDevices in alpha
+   release
+
+## Repos
+
+kubevirt/kubevirt
+
+## Design
+
+For allowing users to consume DRA devices, there are two main changes needed:
+
+1. API changes and the plumbing required in KubeVirt to generate the domain xml with the devices.
+2. Driver Implementation to set the required attributes for KubeVirt to use
+
+This design document focuses on part 1 of the problem. We're introducing a new feature gate `DRADevices`.
+All the API changes will be gated behind this feature gate so as not to break existing functionality.
+
+### API Changes
+
+```go
+type VirtualMachineInstanceSpec struct {
+	..
+	..
+	// ResourceClaims defines which ResourceClaims must be allocated
+	// and reserved before the VMI and hence virt-launcher pod is allowed to start. The resources
+	// will be made available to the domain which consume them
+	// by name.
+	//
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate in kubernetes
+	//  https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/
+	//
+	// This field is immutable.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ResourceClaims []k8sv1.PodResourceClaim `json:"resourceClaims,omitempty"`
+}
+
+type GPU struct {
+	// Name of the GPU device as exposed by a device plugin
+	Name string `json:"name"`
+	// DeviceSource is the name of the device provisioned either by device plugins
+	// or by DRA enabled device
+	// DeviceName string `json:"deviceName"`    <-- inlined into DeviceSource
+	DeviceSource      DeviceSource `json:",inline"`
+	VirtualGPUOptions *VGPUOptions `json:"virtualGPUOptions,omitempty"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag string `json:"tag,omitempty"`
+}
+
+type HostDevice struct {
+	Name string               `json:"name"`
+	// DeviceSource is the name of the device provisioned either by device plugins
+	// or by DRA enabled device
+	// DeviceName string `json:"deviceName"`    <-- inlined into DeviceSource
+	DeviceSource DeviceSource `json:",inline"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag string `json:"tag,omitempty"`
+}
+
+type DeviceSource struct {
+	// DeviceName is the name of the device provisioned by device-plugins
+	DeviceName *string `json:"deviceName,omitempty"`
```

Comment:
This needs to be left untouched imho

Reply:
From user perspective this is untouched. From a go client perspective, we are expressing an intent that the source either device-plugins or dra.

What is the preference? Have all the fields in GPU struct without the inling?

Reply:
The volume API has this pattern, where the source of is specified by an inline struct `VolumeSource`, https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/api/core/v1/schema.go#L776, this was the inspiration for introducing `DeviceSource`

Reply:
If it has been verified that moving a field into an inline struct doesn't cause any trouble on upgrade, which sounds like it's the case, with even a precedent (`VolumeSource`), I think we're good!